internal/types/path.go

```diff
@@ -211,6 +211,12 @@ func insertByPath(doc *Document, path Path) error {
 			case *Document:
 				v.Set(insertedPath.Slice()[suffix], must.NotFail(NewDocument()))
 			case *Array:
+				if insertedPath.Slice()[suffix] == "0" {
+					if err := v.Append(must.NotFail(NewDocument())); err == nil {
```

Comment:
This fix is a bit questionable to me. It just does the name thing as document.

Reply:
What if the test case is `bson.D{{"$inc", bson.D{{"v.1.foo", int32(1)}}}},`? How should we behave?

Reply:
Ok, I had a task with a similar problem :)

I suggest this implementation: https://github.com/FerretDB/FerretDB/pull/1795/files#diff-4dba376d417720306aee8c5e84635828f8d8d9cb9dd4890d67a7827f5138c017R268-R275